### PR TITLE
Improve Canonicalize API

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/8724
Follow on to https://github.com/apache/arrow-datafusion/pull/8780

## Rationale for this change
The new canonicalize API added in https://github.com/apache/arrow-datafusion/pull/8780:
1. Has to be explictly called (so existing users of the ExprSimplifer will not get its benefits by default)
2. Has a diferent API than the existing [`ExprSimplifer::with_guarantees`](https://docs.rs/datafusion/latest/datafusion/optimizer/simplify_expressions/struct.ExprSimplifier.html#method.with_guarantees)

## What changes are included in this PR?
1. Add `Simplifier::with_canonicalize` and update code to use it
2. Add documentation and example

## Are these changes tested?
Yes, with new doc example as well as existing code

## Are there any user-facing changes?
Better api + documentation. 

cc @Jefffrey

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:288]: Here is the new api that is part of the ExprSimplifer that mirrors the guarantees

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs
Comment (yyy1000) [inline datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs:99]: What about moving the initialization of simplifier out like below? It looks cleaner. :)
```rust
        let simplifier = match plan {
            // Canonicalize step won't reorder expressions in a Join on clause.
            // The left and right expressions in a Join on clause are not commutative,
            // since the order of the columns must match the order of the children.
            LogicalPlan::Join(_) => ExprSimplifier::new(info).with_canonicalize(false),
            _ => ExprSimplifier::new(info),
        };
        let expr = plan
            .expressions()
            .into_iter()
            .map(|e| {
                // TODO: unify with `rewrite_preserving_name`
                let original_name = e.name_for_alias()?;
                let new_e = simplifier.simplify(e)?;
                new_e.alias_if_changed(original_name)
            })
            .collect::<Result<Vec<_>>>()?;
```

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (yyy1000) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:271]:
```suggestion
    /// // Expression: a = c AND 1 = b
```
A small typo

**Review by yyy1000 — CHANGES_REQUESTED** [review]
It's a good design! 😃
I helped review it and here are some of my thoughts.

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs
Comment (alamb) [reply datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs]: 😍  I love it. I did it in e7095cdb8 and it looks much nicer now. Thank you @yyy1000

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs:98]: Actually, I just took a look at the join node:

https://github.com/apache/arrow-datafusion/blob/94a6192f6be30b7f6d009bc936a866bf5dcb280c/datafusion/expr/src/logical_plan/plan.rs#L2596-L2599

Was it actually a problem with the `filter` clause and not the `on` clause? :thinking: 

Because the `on` here is a vec of tuples to represent left and right (I was assuming it was a single Expr before)

**Review by Jefffrey — APPROVED** [review]
Looks clean :+1:

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (yyy1000) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:278]: Aha, that's my previous fault.

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs
Comment (yyy1000) [reply datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs:98]: I think it's the problem with `on`.  🤔
In my previous experience, it's a problem when the tuples of `on` keeps as before when optimizing, but the `left` and `right` of `Join` change, so they can't match. 
Not pretty sure.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (alamb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:278]: fixed in 1bb7d09ad

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs
Comment (alamb) [reply datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs]: I agree it is not entirely clear what is wrong with canonicalizing the expressions for joins. I'll add a note to the comment explaining that we don't really understand the issue (maybe someone can look into it as a follow on)

**Review by yyy1000 — APPROVED** [review]
Looks great!

**Comment by alamb** [comment]
> Looks great!

Thanks for the review @yyy1000  -- I am now just waiting for another committer to approve the PR and I'll merge it

**Comment by alamb** [comment]
@Jefffrey  is now a committer (so thank you for your review @Jefffrey 🙏 )

**Pull request merged**
